### PR TITLE
Dev/toddgrun/request concurrency

### PR DESCRIPTION
**Pull request by ryanbrandenburg**
This is a descendent of https://github.com/dotnet/roslyn/pull/66727 to try to take some work on the CLaSP core off of Todd's plate. @ToddGrun is welcome to cherry-pick my commits off here, rework them or we came merge this one instead, whatever works in the end.

Mostly what I've done is just to add a unit test as [requested by DavidB](https://github.com/dotnet/roslyn/pull/66727#discussion_r1099348559).

### CODE REVIEW

**Comment by ToddGrun** [comment]
I'd vote we just move forward with this one. The other is a bit of a mess.

**Comment by ryanbrandenburg** [comment]
CC @dibarbet, @CyrusNajmabadi, @jasonmalinowski, and @davidwengier. Since they reviewed the parent of this PR.

**Comment by ryanbrandenburg** [comment]
My previous strategy for the `ExecuteAsync_WithCancelInProgressWork_CancelsInProgressWorkWhenMutatingRequestArrives` test won't work because we wrap the tasks in a Task.Run. So we're guaranteed that the handler will be exited, but not that the Task returned by requestExecutionQueue.ExecuteAsync is Completed or cancelled. So I changed it to a negative test. IE, I added a debug-assert which will fail if there are ever still items in the "InProgress" list when we try to start the mutating request. Not my favorite strategy, but the alternative seemed to be to re-work the code purely for the sake of this test, which seemed worse.

**Comment by ToddGrun** [comment]
> CC @dibarbet, @CyrusNajmabadi, @jasonmalinowski, and @davidwengier. Since they reviewed the parent of this PR.

I've abandoned the old PR in favor of this one, can y'all take a look at this one sometime early this week? If I'm going to have any chance of getting the work done on my side in preview 2, I'll need this available fairly soon.

**Comment by ToddGrun** [comment]
@dibarbet @CyrusNajmabadi @jasonmalinowski and @davidwengier 

I've abandoned the old PR in favor of this one, can y'all take a look at this one sometime early this week? If I'm going to have any chance of getting the work done on my side in preview 2, I'll need this available fairly soon.

**Review by davidwengier — APPROVED** [review]

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs [hunk 1]
```diff
@@ -186,27 +210,69 @@ private async Task ProcessQueueAsync()
...
+                    // Explicitly ignore this exception as cancellation occurred as a result of our linked cancellation token.
                     // This means either the queue is shutting down or the request itself was cancelled.
                     //   1.  If the queue is shutting down, then while loop will exit before the next iteration since it checks for cancellation.
                     //   2.  Request cancellations are normal so no need to report anything there.
                 }
+                catch (ObjectDisposedException)
+                {
+                    // Explicitly ignore this exception as this can occur during the CreateLinkTokenSource call, and means one of the
+                    // linked cancellationTokens has been cancelled.
+                }
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs:275]: i feel like this needs to move to the place where we make the linked token.  i think we should handle this clearly and specify the exact semantics of what that means at that location.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TaskExtensions.cs [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+
+using System;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Microsoft.CommonLanguageServerProtocol.Framework
+{
+    internal static partial class TaskExtensions
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TaskExtensions.cs:13]: is it not possible to link in this file?  i worry about the code duplication here.

Comment (ToddGrun) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TaskExtensions.cs:13]: I had difficulties doing so. In the original version of this file, there is use of the Contract.ThrowIfNull helper in a couple places. Pulling in that dependency, ended up pulling in multiple other source files which I had originally in the other PR and caused other problems (specifically build errors around multiply defined NullabilityAttributes).

I should point out that this assembly might not be long-lived for the Roslyn repository and has intentionally taken minimal dependencies. NoThrowAwaitable seemed useful enough to warrant the code duplication, whereas all the Contract classes and the resulting dependencies did not.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs (hunk 1 above)
Comment (ToddGrun) [reply src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs:275]: I debated on this when I made the change. The two alternates that I saw utilizing a tighter catch around the CreateLinkedTokenSource call were to catch and then either throw an OperationCanceledException or to catch and then do a continue to skip the rest of that iteration. I'll try out the latter and see if you like it better.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs [hunk 3]
```diff
@@ -186,23 +220,60 @@ private async Task ProcessQueueAsync()
...
+
+                            _ = currentWorkTask.ContinueWith(t =>
+                            {
+                                if (!concurrentlyExecutingTasks.TryRemove(t, out var concurrentlyExecutingTaskCts))
+                                {
+                                    throw new InvalidOperationException($"unexpected failure to remove task from {nameof(concurrentlyExecutingTasks)}");
+                                }
+
+                                concurrentlyExecutingTaskCts.Dispose();
+                            }, CancellationToken.None, TaskContinuationOptions.ExecuteSynchronously, TaskScheduler.Default);
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/RequestExecutionQueue.cs:270]: nit: doc that CT.None here is very intentional as this code is absolutely necessary to run for clearnup *regardless* of how the task completed, and regardless if things were cancelled.

File: src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TaskExtensions.cs [hunk 4]
```diff
@@ -0,0 +1,138 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
```
Comment (CyrusNajmabadi) [inline src/Features/LanguageServer/Microsoft.CommonLanguageServerProtocol.Framework/TaskExtensions.cs:1]: not reviewing.  trusting it was copied minimally.

**Review by CyrusNajmabadi — APPROVED** [review]

**Pull request merged**